src/delta.c

```diff
@@ -38,86 +38,164 @@
 #include "swupd.h"
 #include "xattrs.h"
 
-static void do_delta(char *filename, struct file *file);
-
-void try_delta(struct file *file)
+static bool compute_hash_from_file(char *filename, char *hash)
 {
-	char *filename;
-	struct stat stat;
+	/* TODO: implement this without the indirection of creating a file... */
+	struct file f;
 
-	if (file->is_file == 0) {
-		return;
-	}
+	f.filename = filename;
+	f.use_xattrs = true;
 
-	if (file->deltapeer == NULL) {
-		return;
-	}
-
-	if (file->deltapeer->is_file == 0) {
-		return;
+	populate_file_struct(&f, filename);
+	if (compute_hash(&f, filename) != 0) {
+		return false;
 	}
 
-	if (file->deltapeer->is_deleted) {
-		return;
-	}
-
-	/* check if the full file is there already, because if it is, don't do the delta */
-	string_or_die(&filename, "%s/staged/%s", state_dir, file->hash);
-	if (lstat(filename, &stat) == 0) {
-		free_string(&filename);
-		return;
-	}
-	do_delta(filename, file);
-	free_string(&filename);
+	hash_assign(&f.hash[0], hash);
+	return true;
 }
 
-static void do_delta(char *filename, struct file *file)
+static bool apply_one_delta(char *from_file, char *to_staged, char *delta_file, char *to_hash)
 {
-	char *origin;
-	char *dir, *base, *tmp = NULL, *tmp2 = NULL;
-	char *deltafile = NULL;
 	int ret;
-	struct stat sb;
 
-	string_or_die(&deltafile, "%s/delta/%i-%i-%s-%s", state_dir,
-		      file->deltapeer->last_change, file->last_change, file->deltapeer->hash, file->hash);
+	char *prefixed_from_file;
+	string_or_die(&prefixed_from_file, "%s/%s", path_prefix, from_file);
 
-	ret = stat(deltafile, &sb);
-	if (ret != 0) {
-		free_string(&deltafile);
-		return;
+	ret = apply_bsdiff_delta(prefixed_from_file, to_staged, delta_file);
+	if (ret) {
+		fprintf(stderr, "apply bsdiff delta failed\n");
+		goto fail;
 	}
 
-	tmp = strdup(file->deltapeer->filename);
-	tmp2 = strdup(file->deltapeer->filename);
-
-	dir = dirname(tmp);
-	base = basename(tmp2);
+	xattrs_copy(prefixed_from_file, to_staged);
 
-	string_or_die(&origin, "%s/%s/%s", path_prefix, dir, base);
-
-	ret = apply_bsdiff_delta(origin, filename, deltafile);
-	if (ret) {
-		unlink_all_staged_content(file);
-		goto out;
+	/* TODO: Remove this and rely only on hash verification only? */
```

Comment:
minor: you duplicate `only` in this comment

Reply:
Done.